Stores/src/main/java/AppUpdateWorker.kt

```diff
@@ -0,0 +1,94 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.infomaniak.lib.stores
+
+import android.content.Context
+import androidx.concurrent.futures.CallbackToFutureAdapter
+import androidx.work.*
+import com.google.common.util.concurrent.ListenableFuture
+import com.infomaniak.lib.core.utils.SentryLog
+import com.infomaniak.lib.stores.updatemanagers.WorkerUpdateManager
+import io.sentry.Sentry
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import java.util.concurrent.TimeUnit
+
+class AppUpdateWorker(appContext: Context, params: WorkerParameters) : ListenableWorker(appContext, params) {
+
+    private val updateManager by lazy { WorkerUpdateManager(appContext) }
+
+    override fun startWork(): ListenableFuture<Result> {
+        SentryLog.i(TAG, "Work started")
+
+        return CallbackToFutureAdapter.getFuture { completer ->
+            updateManager.installDownloadedUpdate(
+                onInstallSuccess = { completer.setResult(Result.success()) },
+                onInstallFailure = { exception ->
+                    // This avoid the user being instantly reprompted to download update
+                    Sentry.captureException(exception)
+                    completer.setResult(Result.failure())
+                },
+            )
+        }
+    }
+
+    private fun CallbackToFutureAdapter.Completer<Result>.setResult(result: Result) {
+        set(result)
+        SentryLog.d(TAG, "Work finished")
+    }
+
+    class Scheduler(appContext: Context) {
+
+        private val workManager = WorkManager.getInstance(appContext)
+        private val storesLocalSettings = StoresLocalSettings.getInstance(appContext)
+
+        fun scheduleWorkIfNeeded() {
+            if (BuildConfig.HAS_GPLAY_SERVICES && storesLocalSettings.hasAppUpdateDownloaded) {
+                SentryLog.d(TAG, "Work scheduled")
+
+                val workRequest = OneTimeWorkRequestBuilder<AppUpdateWorker>()
+                    .setConstraints(Constraints.Builder().setRequiresBatteryNotLow(true).build())
+                    // We start with a delayed duration, so that when the app quickly come back to foreground because the user
+                    // was just switching apps, the service is not launched
+                    .setInitialDelay(INITIAL_DELAY, TimeUnit.SECONDS)
+                    .build()
+
+                workManager.enqueueUniqueWork(TAG, ExistingWorkPolicy.KEEP, workRequest)
+            }
+        }
+
+        suspend fun cancelWorkIfNeeded() = withContext(Dispatchers.IO) {
+
+            val workInfo = workManager.getWorkInfos(
+                WorkQuery.Builder.fromUniqueWorkNames(listOf(TAG))
+                    .addStates(listOf(WorkInfo.State.BLOCKED, WorkInfo.State.ENQUEUED))
+                    .build(),
+            ).get()
+
+            workInfo.forEach {
```

Comment:
Is there any case where you have multiple `workInfo` here ? Since your `ExistingWorkPolicy` is `ExistingWorkPolicy.KEEP`, you should only have one right?

Reply:
I suppose, I just used the loop because it's safer than using a single() or first() and is not performance-hungry if there's only one workInfos
If you assure me that there always be only one, I will have no probleme using single()

Reply:
After discussions with @sirambd, we'll keep the forEach in order to be safe, we're not sure it will always have only one work with this query

Reply:
For me, if you have multiple `workInfos`, something is wrong. Since we're using a `KEEP` for the worker, I don't see any case we can have multiple `workInfos`. Maybe we can test multiple time just to see if we can encounter that case or we can add a Sentry to check that and if after some time, we never have multiple `workInfos`, we can remove that ?